appserver/common/container-common/src/main/java/com/sun/enterprise/container/common/impl/util/ClusteredSingletonLookupImplBase.java

```diff
@@ -0,0 +1,117 @@
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package com.sun.enterprise.container.common.impl.util;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IAtomicLong;
+import com.hazelcast.core.ILock;
+import com.hazelcast.core.IMap;
+import com.sun.enterprise.container.common.spi.ClusteredSingletonLookup;
+import com.sun.enterprise.container.common.spi.ClusteredSingletonLookup.SingletonType;
+import fish.payara.nucleus.hazelcast.HazelcastCore;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.glassfish.internal.api.Globals;
+
+/**
+ * Base class for implementing Clustered Singleton Lookups
+ *
+ * @author lprimak
+ */
+@RequiredArgsConstructor
+public abstract class ClusteredSingletonLookupImplBase implements ClusteredSingletonLookup {
+    @Override
+    public ILock getDistributedLock() {
+        return getHazelcastInstance().getLock(getLockKey());
+    }
+
+    @Override
+    public IMap<String, Object> getClusteredSingletonMap() {
+        return getHazelcastInstance().getMap(getMapKey());
+    }
+
+    @Override
+    public  IAtomicLong getClusteredUsageCount() {
+        return getHazelcastInstance().getAtomicLong(getSessionHzKey()+ "/count");
+    }
+
+    private HazelcastInstance getHazelcastInstance() {
+        if (!hzCore.isEnabled()) {
+            throw new IllegalStateException("ClusteredSingleton.getHazelcastInstance() - Hazelcast is Disabled");
+        }
+        return hzCore.getInstance();
+    }
+
+    @Override
+    public boolean isClusteredEnabled() {
+        return hzCore.isEnabled();
+    }
+
+    @Override
+    public boolean isDistributedLockEnabled() {
+        return isClusteredEnabled();
+    }
+    
+    private String makeKeyPrefix() {
+        return String.format("Payara/%s/singleton/", singletonType.name().toLowerCase());
+    }
+
+    private String makeMapKey() {
+        return getKeyPrefix() + componentId;
+    }
+
+    private String makeLockKey() {
+        return getSessionHzKey() + "/lock";
+    }
+
+    private String makeSessionHzKey() {
+        return getKeyPrefix() + componentId + "/" + getClusteredSessionKey();
+    }
+
+
+    private final HazelcastCore hzCore = Globals.getDefaultHabitat().getService(HazelcastCore.class);
+    private final String componentId;
+    private final SingletonType singletonType;
+    private final @Getter(lazy = true, value = AccessLevel.PROTECTED) String keyPrefix = makeKeyPrefix();
+    private final @Getter(lazy = true, value = AccessLevel.PROTECTED) String mapKey = makeMapKey();
+    private final @Getter(lazy = true, value = AccessLevel.PROTECTED) String lockKey = makeLockKey();
+    private final @Getter(lazy = true, value = AccessLevel.PUBLIC) String sessionHzKey = makeSessionHzKey();
```

Comment:
should class variables be declared at the top of a file, rather than the bottom

Reply:
There is a reason I do this... we can take this offline, but I usually put the most important code on top, so it's clear what the file does, and not the variables (implementation detail) and not usually the most important thing in the code